### ISSUE TITLE
Pytest delays when AppMap is installed into a project

### ISSUE THREAD

**Issue by petecheslock** [getappmap/appmap-python#260]
In the fork of [Django-Oscar](https://github.com/land-of-apps/django-oscar), running AppMap tests without AppMap installed will complete the full test suite with `pytest` in about 2m.   [Example GitHub Action with the test run](https://github.com/land-of-apps/django-oscar/actions/runs/6577423541/job/17869028867) 

With AppMap added to the project the test suite takes well over 1 hour to complete. 
Version:
```
$ pip show appmap 
Name: appmap
Version: 1.17.0
Summary: Create AppMap files by recording a Python application.
Home-page: https://github.com/applandinc/appmap-python
Author: Alan Potter
Author-email: alan@app.land
License: MIT
Location: /Users/petecheslock/repos/django-oscar/venv/lib/python3.9/site-packages
Requires: importlib-metadata, importlib-resources, inflection, packaging, PyYAML
Required-by: 
```

Below is a work in progress of which tests are the biggest offenders.

```
tests/integration/catalogue/test_category.py::TestMovingACategory::test_fix_tree
tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_all_categories
tests/integration/catalogue/test_category.py::TestCategoryFactory::test_can_create_multiple_categories
tests/integration/catalogue/test_category.py::TestCategoryFactory::test_updating_subtree_slugs_when_moving_category_to_new_parent
tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_categories_parent
tests/integration/catalogue/test_category.py::TestMovingACategory::test_updates_subtree_names
```

To recreate:
Clone the `appmap` branch of [django-oscar](https://github.com/land-of-apps/django-oscar)

Ensure docker desktop is running.

```
docker compose up -d postgres

python -m venv venv
source venv/bin/activate
python -m pip install --upgrade pip
pip install -e .\[test\]
pip install -r requirements.txt

export DATABASE_PORT=5432
export DATABASE_HOST=localhost
export DATABASE_PASSWORD=postgres
export DATABASE_USER=postgres

# Run Tests
pytest
```

**Comment by apotterri** [comment]
@petecheslock for the sake of posterity, can you update this with the actual version of `appmap-python` that's installed in the project, please? `pip show appmap` should do it.

**Comment by petecheslock** [comment]
> @petecheslock for the sake of posterity, can you update this with the actual version of `appmap-python` that's installed in the project, please? `pip show appmap` should do it.

Whoops! Just updated the comment with the _actual_ version of the code.

**Comment by kgilpin** [comment]
My other question is what’s the elapsed time and size of the AppMaps.

**Comment by petecheslock** [comment]
More details of the affected tests
```
1093.93s call     tests/integration/catalogue/test_category.py::TestCategoryFactory::test_updating_subtree_slugs_when_moving_category_to_new_parent
737.44s call     tests/integration/catalogue/test_category.py::TestMovingACategory::test_updates_subtree_names
698.53s call     tests/integration/catalogue/test_category.py::TestMovingACategory::test_fix_tree
655.09s call     tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_all_categories
603.80s call     tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_categories_parent
215.99s call     tests/integration/catalogue/test_category.py::TestCategoryFactory::test_updating_subtree_when_moving_category_to_new_sibling
140.39s call     tests/integration/catalogue/test_category.py::TestCategoryFactory::test_can_create_multiple_categories
64.49s call     tests/integration/catalogue/test_category.py::TestCategoryFactory::test_can_use_alternative_separator
28.29s call     tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_categories_depth_parent
17.49s call     tests/integration/catalogue/test_category.py::TestMovingACategory::test_updates_instance_name
7.47s setup    tests/integration/catalogue/test_category.py::TestCategory::test_has_children_method
2.71s call     tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_categories_depth
2.53s call     tests/integration/catalogue/test_category.py::TestCategoryFactory::test_can_create_parent_and_child_categories
1.05s call     tests/integration/catalogue/test_category.py::TestCategory::test_supplied_slug_is_not_altered
0.93s call     tests/integration/catalogue/test_category.py::TestCategory::test_unicode_slug
0.62s call     tests/integration/catalogue/test_category.py::TestCategory::test_url_caching
0.58s call     tests/integration/catalogue/test_category.py::TestCategory::test_includes_parents_name_in_full_name
0.19s call     tests/integration/catalogue/test_category.py::TestCategory::test_has_children_method
0.16s teardown tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_categories_parent
0.07s call     tests/integration/catalogue/test_category.py::TestCategoryFactory::test_can_create_single_level_category
```

**Comment by petecheslock** [comment]
<img width="1358" alt="Screenshot 2023-10-19 at 6 14 23 PM" src="https://github.com/getappmap/appmap-python/assets/511733/9b0ed5df-2468-4dde-b858-1794bc737432">
Size of AppMaps with AppMap enabled.

**Comment by petecheslock** [comment]
Timings of tests with appmap not installed

```
-- Docs: https://docs.pytest.org/en/stable/how-to/capture-warnings.html
===================================================================== slowest 20 durations ======================================================================
6.56s setup    tests/integration/catalogue/test_category.py::TestCategory::test_has_children_method
0.07s call     tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_categories_depth_parent
0.06s call     tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_categories_parent
0.06s call     tests/integration/catalogue/test_category.py::TestCategory::test_unicode_slug
0.06s call     tests/integration/catalogue/test_category.py::TestMovingACategory::test_updates_instance_name
0.05s call     tests/integration/catalogue/test_category.py::TestMovingACategory::test_updates_subtree_names
0.05s call     tests/integration/catalogue/test_category.py::TestMovingACategory::test_fix_tree
0.05s call     tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_categories_depth
0.05s call     tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_all_categories
0.05s call     tests/integration/catalogue/test_category.py::TestCategoryFactory::test_updating_subtree_slugs_when_moving_category_to_new_parent
0.05s call     tests/integration/catalogue/test_category.py::TestCategoryFactory::test_updating_subtree_when_moving_category_to_new_sibling
0.04s teardown tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_categories_parent
0.02s call     tests/integration/catalogue/test_category.py::TestCategoryFactory::test_can_create_multiple_categories
0.02s call     tests/integration/catalogue/test_category.py::TestCategory::test_supplied_slug_is_not_altered
0.02s call     tests/integration/catalogue/test_category.py::TestCategoryFactory::test_can_use_alternative_separator
0.02s call     tests/integration/catalogue/test_category.py::TestCategory::test_url_caching
0.02s setup    tests/integration/catalogue/test_category.py::TestMovingACategory::test_fix_tree
0.01s setup    tests/integration/catalogue/test_category.py::TestCategoryTemplateTags::test_all_categories
0.01s call     tests/integration/catalogue/test_category.py::TestCategory::test_slugs_were_autogenerated
0.01s call     tests/integration/catalogue/test_category.py::TestCategory::test_has_children_method
```

**Comment by kgilpin** [comment]
1.4 GB LOL

Can you try the `appmap stats` command?

**Comment by kgilpin** [comment]
Which is see is missing from https://appmap.io/docs/reference/appmap-client-cli.html

the folder version is described here - https://appmap.io/docs/guides/refine-appmaps.html

there is also a single file option which is good too

@ahtrotta cc

**Comment by petecheslock** [comment]
Ok - this seems to be like a good test to focus on for optimization.  

```
 pytest -svv tests/integration/catalogue/test_category.py   
```

This one completed in:

```
19 passed, 3 warnings in 4277.26s (1:11:17)
```

<details>

<summary>The output of `appmap stats tmp/appmap` on only those maps created from this test:</summary>

```
✔ Computing AppMap stats...

Largest AppMaps (which are bigger than 1024kb)
1431MB pytest/tests_integration_catalogue_test_category_TestCategoryFactory_test_updating_subtree_slugs_when_moving_category_to_new_parent.appmap.json
1074.9MB pytest/tests_integration_catalogue_test_category_TestMovingACategory_test_updates_subtree_names.appmap.json
1074MB pytest/tests_integration_catalogue_test_category_TestCategoryTemplateTags_test_all_categories.appmap.json
1073.3MB pytest/tests_integration_catalogue_test_category_TestMovingACategory_test_fix_tree.appmap.json
1009.3MB pytest/tests_integration_catalogue_test_category_TestCategoryTemplateTags_test_categories_parent.appmap.json
366.8MB pytest/tests_integration_catalogue_test_category_TestCategoryFactory_test_updating_subtree_when_moving_category_to_new_sibling.appmap.json
185.2MB pytest/tests_integration_catalogue_test_category_TestCategoryFactory_test_can_create_multiple_categories.appmap.json
92.7MB pytest/tests_integration_catalogue_test_category_TestCategoryFactory_test_can_use_alternative_separator.appmap.json
57.9MB pytest/tests_integration_catalogue_test_category_TestCategoryTemplateTags_test_categories_depth_parent.appmap.json
42MB pytest/tests_integration_catalogue_test_category_TestMovingACategory_test_updates_instance_name.appmap.json
```  


</details>

The `⠇ Computing functions with highest AppMap overhead...` is still running and will post after it finishes.

**Comment by kgilpin** [comment]
You can run the stats command with one argument which is the AppMap to investigate, and you’ll get a lot more detail in that mode.

**Comment by petecheslock** [comment]
Hmm - tried that but getting an error, seems to really want the directory

```
appmap stats tmp/appmap/pytest/tests_integration_catalogue_test_category_TestCategoryFactory_test_updating_subtree_slugs_when_moving_category_to_new_parent.appmap.json
Error: ENOTDIR: not a directory, chdir '/Users/petecheslock/repos/django-oscar' -> 'tmp/appmap/pytest/tests_integration_catalogue_test_category_TestCategoryFactory_test_updating_subtree_slugs_when_moving_category_to_new_parent.appmap.json'
    at wrappedChdir (node:internal/bootstrap/switches/does_own_process_state:112:14)
    at process.chdir (node:internal/worker:99:5)
    at process.<anonymous> (/snapshot/appmap-js/node_modules/graceful-fs/polyfills.js:22:11)
    at process.<anonymous> (/snapshot/appmap-js/node_modules/fs-extra/node_modules/graceful-fs/polyfills.js:22:11)
    at process.chdir (/snapshot/appmap-js/node_modules/jsonfile/node_modules/graceful-fs/polyfills.js:22:11)
    at handleWorkingDirectory (/snapshot/appmap-js/packages/cli/built/lib/handleWorkingDirectory.js:6:17)
    at Object.handler (/snapshot/appmap-js/packages/cli/built/cmds/stats/stats.js:72:57)
    at /snapshot/appmap-js/node_modules/yargs/build/index.cjs:1:9114
    at j (/snapshot/appmap-js/node_modules/yargs/build/index.cjs:1:4993)
    at M.applyMiddlewareAndGetResult (/snapshot/appmap-js/node_modules/yargs/build/index.cjs:1:9083) {
  errno: -20,
  code: 'ENOTDIR',
  syscall: 'chdir',
  path: '/Users/petecheslock/repos/django-oscar',
  dest: 'tmp/appmap/pytest/tests_integration_catalogue_test_category_TestCategoryFactory_test_updating_subtree_slugs_when_moving_category_to_new_parent.appmap.json'
}
```

my appmap version 

```
$ appmap --version
3.102.1
```

**Comment by kgilpin** [comment]
Run with `--help`

**Comment by kgilpin** [comment]
cc @ahtrotta because he wrote the single file option

**Comment by petecheslock** [comment]
Here we go

```
appmap stats -a  tmp/appmap/pytest/tests_integration_catalogue_test_category_TestCategoryFactory_test_updating_subtree_slugs_when_moving_category_to_new_parent.appmap.json
Analyzing AppMap: /Users/petecheslock/repos/django-oscar/tmp/appmap/pytest/tests_integration_catalogue_test_category_TestCategoryFactory_test_updating_subtree_slugs_when_moving_category_to_new_parent.appmap.json

1. function:oscar/apps/catalogue/abstract_models/AbstractCategory#get_ancestors_and_self
      count: 1948178
      estimated size: 722.2 MB

2. function:oscar/apps/catalogue/categories.create_from_sequence
      count: 13
      estimated size: 4.9 KB

3. function:oscar/apps/catalogue/abstract_models/AbstractCategory#save
      count: 8
      estimated size: 4.5 KB

4. function:oscar/apps/catalogue/abstract_models/AbstractCategory#generate_slug
      count: 8
      estimated size: 2.9 KB

5. function:oscar/core/utils.slugify
      count: 8
      estimated size: 2.1 KB

6. function:oscar/core/utils.default_slugifier
      count: 8
      estimated size: 3.0 KB

7. function:oscar/core/utils.cautious_slugify
      count: 8
      estimated size: 2.2 KB

8. function:oscar/apps/catalogue/abstract_models/AbstractCategory#set_ancestors_are_public
      count: 8
      estimated size: 3.0 KB

9. function:oscar/apps/catalogue/abstract_models/AbstractCategory#get_descendants_and_self
      count: 8
      estimated size: 3.0 KB

10. function:oscar/apps/catalogue/abstract_models/ReverseStartsWith#process_lhs
      count: 8
      estimated size: 7.8 KB
```

**Comment by kgilpin** [comment]
cc @apotterri check this out

**Comment by petecheslock** [comment]
Similar story in this case for other maps as well.  

```
appmap stats -a tmp/appmap/pytest/tests_integration_catalogue_test_category_TestMovingACategory_test_updates_subtree_names.appmap.json
Analyzing AppMap: /Users/petecheslock/repos/django-oscar/tmp/appmap/pytest/tests_integration_catalogue_test_category_TestMovingACategory_test_updates_subtree_names.appmap.json

1. function:oscar/apps/catalogue/abstract_models/AbstractCategory#get_ancestors_and_self
      count: 1441314
      estimated size: 540.3 MB

2. function:tests/integration/catalogue/test_category/TestMovingACategory#test_updates_subtree_names
      count: 1
      estimated size: 486.0 bytes
```

**Comment by kgilpin** [comment]
Did you try updating the appmap.yml to exclude this and re-running?

**Comment by petecheslock** [comment]
Sorry - missed this @kgilpin - going to try to set my appmap.yml to this:

```yaml
appmap_dir: tmp/appmap
language: python
name: django-oscar
packages:
  - path: sandbox
  - path: oscar
    exclude:
    - AbstractCategory#get_ancestors_and_self
  - path: tests
  - dist: Django
    exclude:
    - django.db
  ```

**Comment by petecheslock** [comment]
Not entirely sure i'm excluding this function correctly, but I've tried this also and doesn't seem to change things.  Any ideas @kgilpin? 

```yaml
appmap_dir: tmp/appmap
language: python
name: django-oscar
packages:
  - path: sandbox
  - path: oscar
    exclude:
    - function:oscar/apps/catalogue/abstract_models/AbstractCategory#get_ancestors_and_self
  - path: tests
  - dist: Django
    exclude:
    - django.db
```

**Comment by kgilpin** [comment]
@petecheslock check the syntax in the config reference. I don’t think the way you’ve got it here with the `function:` prefix is correct.

**Comment by petecheslock** [comment]
Hmm - can't seem to find the right setting.  Here's the stats view when opening the map.  
<img width="1069" alt="Screenshot 2023-10-24 at 4 07 39 PM" src="https://github.com/getappmap/appmap-python/assets/511733/3d7dd413-2b70-4dc6-981b-20a7a2acab96">


The top offender `oscar/apps/catalogue/abstract_models/AbstractCategory#get_ancestors_and_self`

My most recent attempts in the `appmap.yml`

```yaml
appmap_dir: tmp/appmap
language: python
name: django-oscar
packages:
  - path: sandbox
  - path: oscar
    exclude:
    - AbstractCategory#get_ancestors_and_self
    - AbstractCategory
    - oscar/apps/catalogue/abstract_models
    - apps/catalogue/abstract_models
  - path: tests
  - dist: Django
    exclude:
    - django.db
```

**Comment by kgilpin** [comment]
In Ruby there is a log setting that will show you what's going on with the agent.

@dustinbyrne @apotterri how does Pete exclude this function?

**Comment by petecheslock** [comment]
Thanks -  did find `APPMAP_LOG_LEVEL=info` which i'm running now to see if i can spot anything.

**Comment by petecheslock** [comment]
Latest update to conform with how Python expects.  Saving for posterity. 

```yaml
appmap_dir: tmp/appmap
language: python
name: django-oscar
packages:
  - path: sandbox
  - path: oscar.apps.catalogue.abstract_models
    exclude: 
    - AbstractCategory.get_ancestors_and_self
  - path: tests
  - dist: Django
    exclude:
    - django.db
```

**Comment by petecheslock** [comment]
This issue will be resolved by #263 

Using this appmap.yml will properly record the oscar packages while excluding the one particularly noisy function.

```
appmap_dir: tmp/appmap
language: python
name: django-oscar
packages:
  - path: oscar
    exclude: 
    - apps.catalogue.abstract_models.AbstractCategory.get_ancestors_and_self
```

After setting this the main affected test which used to take about 30 minutes to complete finishes in 16 seconds.
